src/transformers/integrations/awq.py

```diff
@@ -0,0 +1,115 @@
+from ..utils.quantization_config import AwqBackendPackingMethod, AWQLinearVersion
+
+
+if is_torch_available():
+    import torch.nn as nn
+
+if is_accelerate_available():
+    from accelerate import init_empty_weights
+
+
+def replace_with_awq_linear(
+    model,
+    modules_to_not_convert=None,
+    quantization_config=None,
+    current_key_name=None,
+    has_been_replaced=False,
+) -> bool:
+    """
+    Public method that recursively replaces the Linear layers of the given model with AWQ quantized layers.
+    `accelerate` is needed to use this method. Returns the converted model and a boolean that indicates if the
+    conversion has been successfull or not.
+
+    During the module replacement, we also infer the backend to use through the `quantization_config` object.
+
+    Args:
+        model (`torch.nn.Module`):
+            The model to convert, can be any `torch.nn.Module` instance.
+        quantization_config (`AwqConfig`):
+            The quantization config object that contains the quantization parameters.
+        modules_to_not_convert (`list`, `optional`):
+            A list of modules to not convert. If a module name is in the list (e.g. `lm_head`), it will not be
+            converted.
+        current_key_name (`list`, `optional`):
+            A list that contains the current key name. This is used for recursion and should not be passed by the user.
+        has_been_replaced (`bool`, `optional`):
+            A boolean that indicates if the conversion has been successfull or not. This is used for recursion and
+            should not be passed by the user.
+    """
+    if modules_to_not_convert is None:
+        modules_to_not_convert = []
+
+    backend = quantization_config.backend
+
+    if not is_accelerate_available():
+        raise ValueError(
+            "This method requires `accelerate` to be installed. Please install it with `pip install accelerate`"
+        )
+
+    if is_auto_awq_available():
+        if backend == AwqBackendPackingMethod.AUTOAWQ:
+            from awq.modules.linear import WQLinear_GEMM, WQLinear_GEMV
+        elif backend == AwqBackendPackingMethod.LLMAWQ:
+            from awq.quantize.qmodule import WQLinear
+    else:
+        raise ValueError(
+            "AWQ (either `autoawq` or `llmawq`) is not available. Please install it with `pip install autoawq` or check out the installation guide in https://github.com/mit-han-lab/llm-awq"
+        )
+
+    for name, module in model.named_children():
+        if current_key_name is None:
+            current_key_name = []
+        current_key_name.append(name)
+
+        if isinstance(module, nn.Linear) and name not in modules_to_not_convert:
+            # Check if the current key is not in the `modules_to_not_convert`
+            if not any(key in ".".join(current_key_name) for key in modules_to_not_convert):
+                with init_empty_weights():
```

Comment:
It should work without having to use `init_empty_weights` as the whole model is initialized under "init_empty_weights" when we use `device_map`. At least, it worked with gptq on optimum. LMK if it breaks something.

Reply:
ok perfect, removed it!